File: android/measure/src/main/java/sh/measure/android/tracing/TraceSampler.kt
package sh.measure.android.tracing

import sh.measure.android.config.ConfigProvider
import sh.measure.android.utils.Randomizer

internal interface TraceSampler {
    fun shouldSample(): Boolean
}

internal class TraceSamplerImpl(
    private val randomizer: Randomizer,
    private val configProvider: ConfigProvider,
) : TraceSampler {
    override fun shouldSample(): Boolean {
        if (configProvider.traceSamplingRate == 0.0f) {
            return false
        }
        if (configProvider.traceSamplingRate == 1.0f) {
            return true
        }
        return randomizer.random() < configProvider.traceSamplingRate
    }
}

File: flutter/packages/measure_flutter/lib/src/tracing/trace_sampler.dart
import 'package:measure_flutter/src/tracing/randomizer.dart';

import '../config/config_provider.dart';

abstract class TraceSampler {
  bool shouldSample();
}

class TraceSamplerImpl implements TraceSampler {
  final Randomizer _randomizer;
  final ConfigProvider _configProvider;

  TraceSamplerImpl(this._randomizer, this._configProvider);

  @override
  bool shouldSample() {
    if (_configProvider.traceSamplingRate == 0.0) {
      return false;
    }
    if (_configProvider.traceSamplingRate == 1.0) {
      return true;
    }
    return _randomizer.random() < _configProvider.traceSamplingRate;
  }
}
